dlme_airflow/drivers/oai_xml.py

```diff
@@ -62,42 +69,32 @@ def _construct_fields(self, manifest: etree) -> dict:
         return output
 
     def _get_tag(self, el):
-        """Return the element name, after removing namespace and lowercasing."""
-        return re.sub(r"{.+}", "", el.tag).lower()
+        """Return the element name, after removing namespace and lowercasing.
+        If it's a MARC element the tag and code attributes will be used instead.
+        """
+        qel = etree.QName(el)
+        if qel.namespace == NS["marc21"]:
+            if qel.localname in ["controlfield", "datafield"]:
+                return el.attrib["tag"]
+            elif qel.localname == "subfield":
+                return el.attrib["code"]
+            else:
+                return qel.localname
+        else:
+            return qel.localname
 
     # TODO: Discuss if this output shoould be an array (line 63) or a string
     def _from_metadata(self, manifest: etree) -> dict:
-        output = {}
-
-        NS = {
-            "oai_dc": "http://www.openarchives.org/OAI/2.0/oai_dc/",
-            "mods": "http://www.loc.gov/mods/v3",
-        }
-
         if self.metadata_prefix == "oai_dc":
-            oai_block = manifest.xpath("//oai_dc:dc", namespaces=NS)[0]
+            oai_rec = manifest.xpath("//oai_dc:dc", namespaces=NS)[0]
         elif self.metadata_prefix == "mods" or self.metadata_prefix == "mods_no_ocr":
-            oai_block = manifest.xpath("//mods:mods", namespaces=NS)[0]
+            oai_rec = manifest.xpath("//mods:mods", namespaces=NS)[0]
+        elif self.metadata_prefix == "marc21":
+            oai_rec = manifest.xpath("//marc21:record", namespaces=NS)[0]
         else:
             raise Exception(f"Unknown metadata prefix {self.metadata_prefix}")
 
-        for metadata in oai_block.getchildren():
-            tag = self._get_tag(metadata)
-            if tag in output:
-                if isinstance(output[tag], str):
-                    if metadata.text is not None:
-                        output[tag] = [output[tag], metadata.text.strip()]
-                    else:
-                        output[tag] = [output[tag]]
-                else:
-                    output[tag].append(metadata.text.strip())
-            elif metadata.text is not None:
-                output[tag] = metadata.text.strip()
-            # if the element has child elements add them too
-            elif len(metadata) > 0:
-                output.update(self._flatten_tree(metadata))
-
-        return output
+        return self._element_to_dict(oai_rec)
```

Comment:
End ended up extracting this logic into a new method `_element_to_dict()` which is below.